File: app/src/main/java/com/stevenfrew/beatprompter/midi/ArgumentValue.java
package com.stevenfrew.beatprompter.midi;

import com.stevenfrew.beatprompter.BeatPrompterApplication;
import com.stevenfrew.beatprompter.R;

class ArgumentValue extends Value {
    private int mArgumentIndex;
    ArgumentValue(int index)
    {
        mArgumentIndex=index;
    }

    @Override
    byte resolve(byte[] arguments, byte channel) throws ResolutionException {
        if(mArgumentIndex>=arguments.length)
            throw new ResolutionException(BeatPrompterApplication.getResourceString(R.string.not_enough_parameters_supplied));
        return arguments[mArgumentIndex];
    }

    @Override
    boolean matches(Value otherValue) {
        if(otherValue instanceof ArgumentValue)
            return (((ArgumentValue) otherValue).mArgumentIndex==mArgumentIndex);
        return otherValue instanceof WildcardValue;
    }
}

File: app/src/main/java/com/stevenfrew/beatprompter/midi/ArgumentValue.kt
package com.stevenfrew.beatprompter.midi

import com.stevenfrew.beatprompter.BeatPrompterApplication
import com.stevenfrew.beatprompter.R

internal class ArgumentValue(private val mArgumentIndex: Int) : Value() {

    @Throws(ResolutionException::class)
    internal override fun resolve(arguments: ByteArray, channel: Byte): Byte {
        if (mArgumentIndex >= arguments.size)
            throw ResolutionException(BeatPrompterApplication.getResourceString(R.string.not_enough_parameters_supplied))
        return arguments[mArgumentIndex]
    }

    internal override fun matches(otherValue: Value?): Boolean {
        return if (otherValue is ArgumentValue) otherValue.mArgumentIndex == mArgumentIndex else otherValue is WildcardValue
    }
}
